js/anxpro.js

```diff
@@ -623,13 +1090,23 @@ module.exports = class anxpro extends Exchange {
         } else {
             this.checkRequiredCredentials ();
             let nonce = this.nonce ();
-            body = this.urlencode (this.extend ({ 'nonce': nonce }, query));
+            let auth = undefined;
+            let contentType = undefined;
+            if (api === 'v3private') {
+                body = this.json (this.extend ({ 'tonce': nonce * 1000 }, query));
+                const path = url.replace ('https://anxpro.com/', '');
+                auth = path + '\0' + body;
+                contentType = 'application/json';
+            } else {
+                body = this.urlencode (this.extend ({ 'nonce': nonce }, query));
+                // eslint-disable-next-line quotes
+                auth = request + "\0" + body;
+                contentType = 'application/x-www-form-urlencoded';
+            }
             let secret = this.base64ToBinary (this.secret);
-            // eslint-disable-next-line quotes
-            let auth = request + "\0" + body;
             let signature = this.hmac (this.encode (auth), secret, 'sha512', 'base64');
             headers = {
-                'Content-Type': 'application/x-www-form-urlencoded',
+                'Content-Type': contentType,
```

Comment:
Let's not set it in cases when it's undefined, because this is not very portable to other languages and environments that treat `undefined/None/null` values differently. I suggest:

```JavaScript
headers = {
    'Rest-Key': this.apiKey,
    // ...
};
if (contentType !== undefined) {
    headers['Content-Type'] = contentType;
}
```

↑  The same logic is true for defaults (`limit`, `since`) throughout this file – would be great if we could clean this up everywhere. Thx!

Reply:
I don't understand

Reply:
Basically, we should not send a value if it is `undefined`.

Reply:
I see

Reply:
That's a shame, it's a lot of checks.  Have you considered just removing undefined fields when we collapse the parameters?

Reply:
Yes, but first adding an undefined value and then removing it from there – is in general more costly than simply not adding it in the first place.

Reply:
... so instead of checking every single optional parameter....

        if (limit !== undefined)
            request['max'] = limit;

We could allow undefined values and remove them in the sign method (via `extractParams`)

Reply:
> We could allow undefined values and remove them in the sign method

We could, indeed, but doesn't mean we should ) Because removing them in each and every sign() call will be more costly than not adding them under an if-condition.

Reply:
... or how about a utility method, something like 
```

    clean (obj) {
        const keys = Object.keys (obj);
        const clean = {};
        for (let i = 0; i < keys.length; i++) {
            const key = keys[i];
            const value = obj[key];
            if (value === undefined)
                continue;
            clean[key] = value;
        }
        return clean;
    }

    async fetchTransactions (code = undefined, since = undefined, limit = undefined, params = {}) {
        const request = this.clean ({
            'from': since,
            'ccy': code,
            'max': limit,
        });
```

Reply:
> ... or how about a utility method, something like

It's not about _"we don't know how to do this"_, it's about _"we think we should not do this, due to the above reasoning"_ ) I mean running more and more additional loops is really against the best design practices and in general it helps with writing conscious and consistent code. Otherwise we will get tons of outdated non-existent params that will be added and filtered out silently here and there – that's a good substrate for bugs.

Reply:
There is a cost, yes, but it's minimal.  I mean compared with send all this stuff across the internet it's negligible.  I think that the improvement in the readability of the code (many fewer lines, easier to write etc) would be worth it.

Reply:
ok

Reply:
> Because removing them in each and every sign() call will be more costly than not adding them under an if-condition

I'm not sure this is correct as we're iterating over them all anyway.  You're swapping a bunch of if statements in one place for another.

But you're right about being able to send keys without values to some exchanges.

Reply:
You could solve the keys without values problem by specifying a null instead of undefined.

Reply:
> would be worth it.

I don't really think so – because contributors tend to make mistakes under relaxed rules, see the above answer – most contributors will start "forgetting" those params here and there. And you'll have a mix impossible to resolve after a few months of collecting those undefined values that quickly get outdated and "lost" in the code. It's just a harder to maintain, if something breaks down, knowing that you cannot rely on any of those values and have to re-check them all the time. 

In contrast to that we think that the code has to be pedantic, not relaxed.  Making the rules more strict rather than more relaxed helps achieve a higher level of portability in the long run. Just because we could do the relaxed approach – doesn't mean we should do it, I think you're missing the reasoning mentioned here: https://github.com/ccxt/ccxt/pull/5169#discussion_r286390619. 

So, the short answer is: we don't do this because we think doing so is more buggy in general.

Reply:
Anyway, I've added a bunch of parameter checks to anxpro

Reply:
> So, the short answer is: we don't do this because we think doing so is more buggy in general.

Fair enough, that's a perfectly good answer :)